File: softeer/좌석관리.cpp
#include <iostream>
#include <sstream>
#include <climits>
#include <vector>

using namespace std;

int n,m,q;
int dx[] = {1,-1,0,0};
int dy[] = {0,0,1,-1};

int map[21][21];

// 0: 안먹, 1: 자리잡음, 2: 이미 다먹음
int user[10001]; // 상태
pair<int,int> location[10001]; //좌표
vector<pair<int,int>> curSeat; // 앉은 자리 좌표

int getDist(int x,int y){
    int vSize = curSeat.size();
    int minVal = INT_MAX;
    
    for(int i = 0; i < vSize; i++){
        int curDist = (curSeat[i].first - x)*(curSeat[i].first - x) + (curSeat[i].second - y)*(curSeat[i].second - y);
        if(minVal > curDist) minVal = curDist;
    }

    return minVal;
}

bool findSeat(int idx){
    int x = -1, y = -1;

    if(curSeat.size() != 0){
        int maxVal = -1;
        for(int i = 1; i <= n; i++){
            for(int j = 1; j <= m; j++){
                if(map[i][j] > 0) continue;
                int curDist = getDist(i,j);
                if(maxVal < curDist){
                    maxVal = curDist;
                    x = i;
                    y = j;
                }
            }
        }
    }else{
        x = 1, y = 1;
    }


    if( x == -1 && y == -1){
        return false;
    }

    user[idx] = 1;
    curSeat.push_back({x,y});
    location[idx] = {x,y};
    map[x][y]++;

    for(int i = 0; i < 4; i++){
        int nx = x + dx[i];
        int ny = y + dy[i];
        if(nx <= 0 || nx > n || ny <= 0 || ny > m) continue;

        map[nx][ny]++;
    }

    return true;
}

int main(){
    cin >> n >> m >> q;
    cin.ignore(100,'\n');

    for(int i = 0; i < q; i++){
        string inp;
        getline(cin,inp);
        istringstream iss(inp);
        string cmd;
        int id;
        iss >> cmd >> id;

        if(cmd == "In"){
            if(user[id] == 0){
                if(findSeat(id)){
                    cout << id << " gets the seat (" << location[id].first << ", " << location[id].second << ").\n";
                }else{
                    cout << "There are no more seats.\n";
                }
            }else if(user[id] == 1){
                cout << id << " already seated.\n";
            }else if(user[id] == 2){
                cout << id << " already ate lunch.\n";
            }
        }else if(cmd == "Out"){
            if(user[id] == 0){
                cout << id << " didn't eat lunch.\n";
            }else if(user[id] == 1){
                cout << id << " leaves from the seat (" << location[id].first << ", " << location[id].second << ").\n";

                user[id] = 2;
                map[location[id].first][location[id].second]--;
                
                for(int j = 0; j < 4; j++){
                    int nx = location[id].first + dx[j];
                    int ny = location[id].second + dy[j];
                    if(nx <= 0 || nx > n || ny <= 0 || ny > m) continue;
                    map[nx][ny]--;
                }

                int cSize = curSeat.size();
                for(int j = 0; j< cSize; j++){
                    if(curSeat[j].first == location[id].first && curSeat[j].second == location[id].second){
                        curSeat.erase(curSeat.begin() + j);
                        break;
                    }
                }

                location[id] = {0,0};
            }else if(user[id] == 2){
                cout << id << " already left seat.\n";
            }
        }
    }

    return 0;
}

File: softeer/좌석관리.java
import java.util.*;
import java.io.*;

class Main{
    static int n,m,q;

    static int map[][];
    static Pair userLocation[] = new Pair[10001];
    static int userState[] = new int[10001];
    static List<Pair> curSeat = new ArrayList<>();
    static int dx[] = {-1,1,0,0};
    static int dy[] = {0,0,-1,1};

    static class Pair{
        int x;
        int y;
        Pair(int x, int y){
            this.x = x;
            this.y = y;
        }

        @Override
        public boolean equals(Object obj){
            Pair t = (Pair)obj;
            return (this.x == t.x && this.y == t.y);
        }
    }

    static int getDist(int x,int y){
        int cSize = curSeat.size();
        int minVal = Integer.MAX_VALUE;
        for(int i = 0; i < cSize; i++){
            int curDist = (curSeat.get(i).x - x) * (curSeat.get(i).x - x) + (curSeat.get(i).y - y)*(curSeat.get(i).y - y);
            if(minVal > curDist) minVal = curDist;
        }

        return minVal;
    }

    static void setSeat(int x, int y){
        map[x][y]++;
        curSeat.add(new Pair(x,y));

        for(int i = 0; i < 4; i++){
                int nx = x + dx[i];
                int ny = y + dy[i];

                if(nx <= 0 || nx > n || ny <= 0 || ny > m) continue;

                map[nx][ny]++;
        }
    }

    static boolean findSeat(int id){
        int x = -1, y = -1;

        if(curSeat.isEmpty()){
            x = 1;
            y = 1;
        }else{
            int maxVal = Integer.MIN_VALUE;
            for(int i = 1; i <= n; i++){
                for(int j = 1; j <= m; j++){
                    if(map[i][j] > 0) continue;
                    int curDist = getDist(i,j);
                    if(maxVal < curDist){
                        x = i;
                        y = j;
                        maxVal = curDist;
                    }
                }
            }
        }

        if(x == - 1 && y == -1){
            return false;
        }

        setSeat(x, y);
        userLocation[id] = new Pair(x,y);   
        userState[id] = 1;

        return true;
    }

    static void outSeat(int id){
        userState[id] = 2;
        int x = userLocation[id].x;
        int y = userLocation[id].y;
        
        map[x][y]--;
        curSeat.remove(new Pair(x,y));

        for(int i = 0; i < 4; i++){
                int nx = x + dx[i];
                int ny = y + dy[i];

                if(nx <= 0 || nx > n || ny <= 0 || ny > m) continue;

                map[nx][ny]--;
        }

        userLocation[id].x = 0;
        userLocation[id].y = 0;
    }

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
    
        String[] inp = br.readLine().split(" ");
        n = Integer.parseInt(inp[0]);
        m = Integer.parseInt(inp[1]);
        q = Integer.parseInt(inp[2]);

        map = new int[n+1][m+1];

        for(int i = 0; i < q; i++){
            inp = br.readLine().split(" ");
            int id = Integer.parseInt(inp[1]);

            if(inp[0].equals("In")){
                if(userState[id] == 0){
                    if(findSeat(id)){
                        bw.append(id + " gets the seat (" + userLocation[id].x + ", " + userLocation[id].y + ").\n");
                    }else{
                        bw.append("There are no more seats.\n");
                    }
                }else if(userState[id] == 1){
                    bw.append(id + " already seated.\n");
                }else if(userState[id] == 2){
                    bw.append(id + " already ate lunch.\n");
                }
            }else if(inp[0].equals("Out")){
                if(userState[id] == 0){
                    bw.append(id + " didn't eat lunch.\n");
                }else if(userState[id] == 1){
                    bw.append(id + " leaves from the seat ("+ userLocation[id].x + ", " + userLocation[id].y + ").\n");
                    outSeat(id);
                }else if(userState[id] == 2){
                    bw.append(id + " already left seat.\n");
                }
            }
            bw.flush();
        }

        bw.close();
        br.close();
    }
}